components/motor/dimensionengineering/sabertooth.go

```diff
@@ -0,0 +1,539 @@
+			if err != nil {
+				return nil, err
+			}
+			if err := decoder.Decode(attributes); err != nil {
+				return nil, err
+			}
+			return &conf, nil
+		},
+		&Config{},
+	)
+}
+
+func newController(c *Config, logger golog.Logger) (*controller, error) {
+	if c.SerialAddress < 128 || c.SerialAddress > 135 {
+		return nil, errors.New("address is out of range. valid addresses are 127 to 135")
+	}
+	ctrl := new(controller)
+	ctrl.activeAxes = make(map[int]bool)
+	ctrl.serialDevice = c.SerialPath
+	ctrl.logger = logger
+	ctrl.address = c.SerialAddress
+
+	if c.TestChan != nil {
+		ctrl.testChan = c.TestChan
+	} else {
+		serialOptions := serial.OpenOptions{
+			PortName:          c.SerialPath,
+			BaudRate:          uint(c.BaudRate),
+			DataBits:          8,
+			StopBits:          1,
+			MinimumReadSize:   1,
+			RTSCTSFlowControl: true,
+		}
+
+		port, err := serial.Open(serialOptions)
+		if err != nil {
+			return nil, err
+		}
+		ctrl.port = port
+	}
+
+	ctrl.activeAxes[1] = false
+	ctrl.activeAxes[2] = false
+
+	return ctrl, nil
+}
+
+// NewMotor returns a Sabertooth driven motor.
+func NewMotor(ctx context.Context, c *Config, logger golog.Logger) (motor.LocalMotor, error) {
+	globalMu.Lock()
+	ctrl, ok := controllers[c.SerialPath]
+	if !ok {
+		newCtrl, err := newController(c, logger)
+		if err != nil {
+			return nil, err
+		}
+		controllers[c.SerialPath] = newCtrl
+		ctrl = newCtrl
+	}
+	globalMu.Unlock()
+
+	ctrl.mu.Lock()
+	defer ctrl.mu.Unlock()
+
+	// is on a known/supported amplifier only when map entry exists
+	claimed, ok := ctrl.activeAxes[c.MotorChannel]
+	if !ok {
+		return nil, fmt.Errorf("invalid Sabertooth motor axis: %d", c.MotorChannel)
+	}
+	if claimed {
+		return nil, fmt.Errorf("axis %d is already in use", c.MotorChannel)
+	}
+	ctrl.activeAxes[c.MotorChannel] = true
+
+	// We don't need to convert MinPowerPct because unset it will be 0, which is a safe default
+	minPowerPct := c.MinPowerPct
+
+	// Convert the MaxPowerPct to a default of 100, or whatever the user supplies
+	var maxPowerPct float64
+	if c.MaxPowerPct != 0.0 {
+		maxPowerPct = c.MaxPowerPct
+	} else {
+		maxPowerPct = 100.0
+	}
+
+	m := &Motor{
+		c:           ctrl,
+		Channel:     c.MotorChannel,
+		dirFlip:     c.DirectionFlip,
+		minPowerPct: minPowerPct,
+		maxPowerPct: maxPowerPct,
+	}
+
+	if err := m.configure(c); err != nil {
+		return nil, err
+	}
+
+	if c.RampValue > 0 {
+		setRampCmd, err := newCommand(c.SerialAddress, setRamping, c.MotorChannel, byte(c.RampValue))
+		if err != nil {
+			return nil, err
+		}
+
+		err = m.c.sendCmd(setRampCmd)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return m, nil
+}
+
+// IsPowered returns if the motor is currently on or off.
+func (m *Motor) IsPowered(ctx context.Context, extra map[string]interface{}) (bool, error) {
+	return m.isOn, nil
+}
+
+// Close stops the motor and marks the axis inactive.
+func (m *Motor) Close() {
+	m.c.mu.Lock()
+	active := m.c.activeAxes[m.Channel]
+	m.c.mu.Unlock()
```

Comment:
Changing this to a defer actually poses some issues. The `m.Stop` method also locks the same mutex, so it deadlocks with a `defer` here because the mutex isn't unlocked before `m.Stop` attempts to acquire the mutex.